File: server/src/main/java/lighting/server/scene/SceneController.java
package lighting.server.scene;

import java.io.IOException;
import java.util.List;

import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.springframework.web.bind.annotation.CrossOrigin;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PutMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RestController;

import lighting.server.IO.IOService;

@RestController
@CrossOrigin(origins = "http://localhost:4200")
public class SceneController {

	private static final Logger log = LogManager.getLogger(SceneController.class);

	private final SceneService sceneService;
	private final IOService iOService;

	public SceneController(final SceneService sceneService, final IOService iOService) {
		this.sceneService = sceneService;
		this.iOService = iOService;
	}

	@GetMapping(value = "/api/recordSceneSingleFrame/{button_id}")
	public boolean recordSceneSingleFrame(@PathVariable int button_id) {
		log.info("Recorded a scene with a single frame to button " + button_id);
		return sceneService.recordScene(button_id);
	}

	@GetMapping(value = "/api/recordSceneMultipleFrames/{button_id}")
	public boolean recordSceneMultipleFrames(@PathVariable int button_id) {
		log.info("Recorded a scene with multiple frames to button " + button_id);
		return sceneService.recordSceneMultipleFrames(button_id);
	}

	@GetMapping(value = "/api/stopRecording")
	public boolean stopRecording() {
		log.info("Stopped recording");
		return sceneService.stopRecording();
	}

	@PutMapping(value = "/api/savescene/")
	public void saveScene(@RequestBody Scene scene) {
		try {
			iOService.updateSceneFromDisk(scene);
			log.info("Saved/updated scene to disk with ID: " + scene.getId());
		} catch (IOException e) {
			log.error("Could not save scene to disk with ID: " + scene.getId(), e);
		}
	}

	@GetMapping(value = "api/downloadscene/{scene_id}")
	public String downloadScene(@PathVariable String scene_id) {
		try {
			log.info("Downloaded scene with ID: " + scene_id);
			return iOService.downloadScene(scene_id);
		} catch (IOException e) {
			log.error("Could not download scene with ID: " + scene_id, e);
		}
		return null;
	}

	@GetMapping(value = "/api/playscene/{button_id}")
	public boolean playSceneFromButton(@PathVariable int button_id) {
		try {
			log.info("Playing scene from button " + button_id);
			return sceneService.playSceneFromButton(button_id);
		} catch (IOException e) {
			log.error("Could not play scene from button " + button_id, e);
			return false;
		}
	}

	@GetMapping(value = "/api/playscenefromid/{id}")
	public boolean playSceneFromId(@PathVariable String id) {
		try {
			log.info("Playing scene from ID " + id);
			return sceneService.playSceneFromId(id);
		} catch (IOException e) {
			log.error("Could not play scene from ID " + id, e);
			return false;
		}
	}

	@GetMapping(value = "/api/stop")
	public void stop() {
		sceneService.stop();
		log.info("Stopped playing a scene");
	}

	@GetMapping(value = "/api/pause/{bool}")
	public void pause(@PathVariable boolean bool) {
		sceneService.pause(bool);
		log.info("Paused a playing scene");
	}

	@GetMapping(value = "/api/sceneslist")
	public List<Scene> getScenes() {
		try {
			log.info("Retrieved all scenes from disk");
			return iOService.getAllScenesFromDisk();
		} catch (IOException e) {
			log.error("Could not retrieve all scenes from disk", e);
		}
		return null;
	}

	@GetMapping(value = "/api/deletescene/{scene_id}")
	public void deleteScene(@PathVariable String scene_id) {
		try {
			iOService.deleteSceneFromDisk(scene_id);
			log.info("Scene deleted with ID: " + scene_id);
		} catch (IOException e) {
			log.error("Could not delete scene with ID: " + scene_id, e);
		}
	}

	@GetMapping(value = "/api/getscene/{scene_id}")
	public Scene getSceneById(@PathVariable String scene_id) {
		try {
			log.info("Retrieved scene from disk with ID: " + scene_id);
			return iOService.getSceneFromDisk(scene_id);
		} catch (IOException e) {
			log.error("Could not retrieved scene from disk with ID: " + scene_id, e);
		}
		return null;
	}

	@GetMapping(value = "api/getbuttons")
	public List<Boolean> getButtons() {
		try {
			log.info("Retrieved buttons status");
			return iOService.getButtons();
		} catch (IOException e) {
			log.error("Could not retrieve buttons status", e);
		}
		return null;
	}

	@GetMapping(value = "api/getPages")
	public int getPages() {
		try {
			log.info("Retrieved page size");
			return iOService.getSettingsFromDisk().getButtonPages();
		} catch (IOException e) {
			log.error("Could not retrieve page size", e);
		}
		return 0;
	}
}

File: server/src/main/scala/lassi/server/scene/SceneController.scala
package lassi.server.scene

import lassi.domain.Scene
import lassi.log.Log
import lassi.server.repository.SceneRespository
import lassi.server.repository.SettingsRepository
import org.springframework.web.bind.annotation.DeleteMapping
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.PutMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RestController

import java.io.IOException

@RestController
class SceneController(
  sceneService: SceneService,
  sceneRespository: SceneRespository,
  settingsRepository: SettingsRepository
) {

  private val log = Log(classOf[SceneController])

  @GetMapping(value = Array("/api/scenes"))
  def getScenes: Seq[Scene] = {
    try {
      log.infoElapsed {
        val scenes = sceneRespository.getAllScenesFromDisk
        (s"get ${scenes.size} scenes", scenes)
      }
    } catch {
      case e: IOException =>
        log.error("Could not retrieve all scenes from disk", e)
        Seq.empty
    }
  }

  @GetMapping(value = Array("/api/scenes/{sceneId}"))
  def getScene(@PathVariable sceneId: String): Scene = {
    try {
      log.infoElapsed {
        val scene = sceneRespository.getSceneFromDisk(sceneId)
        (s"Get scene $sceneId", scene)
      }
    } catch {
      case e: IOException =>
        log.error("Could not retrieved scene from disk with ID: " + sceneId, e)
        null
    }
  }

  @DeleteMapping(value = Array("/api/scenes/{sceneId}"))
  def deleteScene(@PathVariable sceneId: String): Unit = {
    try {
      log.infoElapsed {
        sceneRespository.deleteSceneFromDisk(sceneId)
        (s"Delete scene $sceneId", ())
      }
    } catch {
      case e: IOException =>
        log.error(s"Could not delete scene $sceneId", e)
    }
  }

  @GetMapping(value = Array("/api/record-scene/{buttonId}"))
  def recordSceneSingleFrame(@PathVariable buttonId: Int): Boolean = {
    log.info(s"Record scene with a single frame to button $buttonId")
    sceneService.recordScene(buttonId)
  }


  @GetMapping(value = Array("/api/recordSceneMultipleFrames/{buttonId}"))
  def recordSceneMultipleFrames(@PathVariable buttonId: Int): Boolean = {
    log.info(s"Record scene with multiple frames to button $buttonId")
    sceneService.recordSceneMultipleFrames(buttonId)
  }

  @GetMapping(value = Array("/api/stopRecording"))
  def stopRecording: Boolean = {
    log.info("Stop recording")
    sceneService.stopRecording
  }

  @PutMapping(value = Array("/api/savescene/"))
  def saveScene(@RequestBody scene: Scene): Unit = {
    try {
      sceneRespository.updateSceneFromDisk(scene)
      log.info(s"Save scene ${scene.id}")
    } catch {
      case e: IOException =>
        log.error(s"Could not save scene ${scene.id}", e)
    }
  }

  @GetMapping(value = Array("/api/playscene/{buttonId}"))
  def playSceneFromButton(@PathVariable buttonId: Int): Boolean = try {
    log.info(s"Play scene from button $buttonId")
    sceneService.playSceneFromButton(buttonId)
  } catch {
    case e: IOException =>
      log.error(s"Could not play scene from button $buttonId", e)
      false
  }

  @GetMapping(value = Array("/api/playscenefromid/{sceneId}"))
  def playSceneFromId(@PathVariable sceneId: String): Boolean = try {
    log.info(s"Playing scene $sceneId")
    sceneService.playSceneFromId(sceneId)
  } catch {
    case e: IOException =>
      log.error("Could not play scene $sceneId", e)
      false
  }

  @GetMapping(value = Array("/api/stop"))
  def stop(): Unit = {
    sceneService.stop()
    log.info("Stop playing scene")
  }

  @GetMapping(value = Array("/api/pause/{bool}"))
  def pause(@PathVariable bool: Boolean): Unit = {
    sceneService.pause(bool)
    log.info("Pause playing scene")
  }

  @GetMapping(value = Array("api/getbuttons"))
  def getButtons: Seq[Boolean] = {
    try {
      log.info("Retrieved buttons status")
      sceneRespository.getButtons
    } catch {
      case e: IOException =>
        log.error("Could not retrieve buttons status", e)
        Seq.empty
    }
  }

  @GetMapping(value = Array("api/getPages"))
  def getPages: Int = {
    try {
      log.info("Retrieved page size")
      settingsRepository.readSettings.buttonPageCount
    } catch {
      case e: IOException =>
        log.error("Could not retrieve page size", e)
        0
    }
  }
}
